src/datafusion/helpers.rs

```diff
@@ -0,0 +1,228 @@
+
+// Checks whether the given expression can be resolved using only the columns `col_names`.
+// Copied from datafusion, because it's not accessible from the outside.
+pub fn expr_applicable_for_cols(col_names: &[String], expr: &Expr) -> bool {
+    let mut is_applicable = true;
+    expr.apply(&mut |expr| match expr {
+        Expr::Column(datafusion_common::Column { ref name, .. }) => {
+            is_applicable &= col_names.contains(name);
+            if is_applicable {
+                Ok(VisitRecursion::Skip)
+            } else {
+                Ok(VisitRecursion::Stop)
+            }
+        }
+        Expr::Literal(_)
+        | Expr::Alias(_)
+        | Expr::OuterReferenceColumn(_, _)
+        | Expr::ScalarVariable(_, _)
+        | Expr::Not(_)
+        | Expr::IsNotNull(_)
+        | Expr::IsNull(_)
+        | Expr::IsTrue(_)
+        | Expr::IsFalse(_)
+        | Expr::IsUnknown(_)
+        | Expr::IsNotTrue(_)
+        | Expr::IsNotFalse(_)
+        | Expr::IsNotUnknown(_)
+        | Expr::Negative(_)
+        | Expr::Cast { .. }
+        | Expr::TryCast { .. }
+        | Expr::BinaryExpr { .. }
+        | Expr::Between { .. }
+        | Expr::Like { .. }
+        | Expr::SimilarTo { .. }
+        | Expr::InList { .. }
+        | Expr::Exists { .. }
+        | Expr::InSubquery(_)
+        | Expr::ScalarSubquery(_)
+        | Expr::GetIndexedField { .. }
+        | Expr::GroupingSet(_)
+        | Expr::Case { .. } => Ok(VisitRecursion::Continue),
+
+        Expr::ScalarFunction(scalar_function) => match &scalar_function.func_def {
+            ScalarFunctionDefinition::BuiltIn(fun) => match fun.volatility() {
+                Volatility::Immutable => Ok(VisitRecursion::Continue),
+                Volatility::Stable | Volatility::Volatile => {
+                    is_applicable = false;
+                    Ok(VisitRecursion::Stop)
+                }
+            },
+            ScalarFunctionDefinition::UDF(fun) => match fun.signature().volatility {
+                Volatility::Immutable => Ok(VisitRecursion::Continue),
+                Volatility::Stable | Volatility::Volatile => {
+                    is_applicable = false;
+                    Ok(VisitRecursion::Stop)
+                }
+            },
+            ScalarFunctionDefinition::Name(_) => {
+                internal_err!("Function `Expr` with name should be resolved.")
+            }
+        },
+
+        Expr::AggregateFunction { .. }
+        | Expr::Sort { .. }
+        | Expr::WindowFunction { .. }
+        | Expr::Wildcard { .. }
+        | Expr::Unnest { .. }
+        | Expr::Placeholder(_) => {
+            is_applicable = false;
+            Ok(VisitRecursion::Stop)
+        }
+    })
+    .unwrap();
+    is_applicable
+}
+
+// Below is all the logic necessary (I think) to convert a PhysicalExpr into a ZarrChunkFilter.
+// The logic is mostly copied from datafusion, and is simplified here for the zarr use case.
+pub struct ZarrFilterCandidate {
```

Comment:
So in the end, this is still mostly index based. I'll make a few comments below to clarify the logic (I should actually write those comments in code, I'll do that before we merge), but column names are basically just used as an intermediate step.